src/couch_file.erl

```diff
@@ -531,27 +537,96 @@ find_header(Fd, Block) ->
+        <<RawBin:TotalSize/binary, _/binary>> = RestBlock;
     true ->
         {ok, Missing} = file:pread(
             Fd, (Block * ?SIZE_BLOCK) + 5 + byte_size(RestBlock),
-            TotalBytes - byte_size(RestBlock)),
+            TotalSize - byte_size(RestBlock)),
         RawBin = <<RestBlock/binary, Missing/binary>>
     end,
     <<Md5Sig:16/binary, HeaderBin/binary>> =
         iolist_to_binary(remove_block_prefixes(5, RawBin)),
     Md5Sig = couch_crypto:hash(md5, HeaderBin),
     {ok, HeaderBin}.
 
+
+%% Read a configurable number of block locations at a time using
+%% file:pread/2.  At each block location, read ?PREFIX_SIZE (5) bytes;
+%% if the first byte is <<1>>, assume it's a header block, and the
+%% next 4 bytes hold the size of the header.
+-spec find_header_pread2(file:fd(), block_id()) ->
+    {ok, binary()} | no_valid_header.
+find_header_pread2(Fd, Block) ->
+    find_header_pread2(Fd, Block, read_count()).
+
+-spec find_header_pread2(file:fd(), block_id(), non_neg_integer()) ->
+    {ok, binary()} | no_valid_header.
+find_header_pread2(_Fd, Block, _ReadCount) when Block < 0 ->
+    no_valid_header;
+find_header_pread2(Fd, Block, ReadCount) ->
+    Locations = block_locations(Block, ReadCount),
+    {ok, DataL} = file:pread(Fd, [{L, ?PREFIX_SIZE} || L <- Locations]),
+    case locate_header(Fd, header_locations(Locations, DataL)) of
+        {ok, _Location, HeaderBin} ->
+            io:format("header at block ~p~n", [_Location div ?SIZE_BLOCK]), % TEMP
+            {ok, HeaderBin};
+        _ ->
+            ok = file:advise(
+                Fd, hd(Locations), ReadCount * ?SIZE_BLOCK, dont_need),
+            NextBlock = hd(Locations) div ?SIZE_BLOCK - 1,
+            find_header_pread2(Fd, NextBlock, ReadCount)
+    end.
+
+-spec read_count() -> non_neg_integer().
+read_count() ->
+    config:get_integer("couchdb", "find_header_read_count", ?DEFAULT_READ_COUNT).
+
+-spec block_locations(block_id(), non_neg_integer()) -> [location()].
+block_locations(Block, ReadCount) ->
+    First = max(0, Block - ReadCount + 1),
+    [?SIZE_BLOCK*B || B <- lists:seq(First, Block)].
+
+-spec header_locations([location()], [data()]) -> [{location(), header_size()}].
+header_locations(Locations, DataL) ->
+    lists:foldl(fun
+        ({Loc, <<1, HeaderSize:32/integer>>}, Acc) ->
+            [{Loc, HeaderSize} | Acc];
+        (_, Acc) ->
+            Acc
+        end, [], lists:zip(Locations, DataL)).
+
+-spec locate_header(file:fd(), [{location(), header_size()}]) ->
+    {ok, binary()} | not_found.
+locate_header(_Fd, []) ->
+    not_found;
+locate_header(Fd, [{Location, Size}|LocationSizes]) ->
+    case (catch load_header(Fd, Location, Size)) of
+        {ok, HeaderBin} ->
+            {ok, Location, HeaderBin};
+        _Error ->
+            locate_header(Fd, LocationSizes)
+    end.
+
+-spec load_header(file:fd(), location(), header_size()) ->
+    {ok, binary()} | no_return().
+load_header(Fd, Location, Size) ->
```

Comment:
How about something like this:

```erlang
load_header(Fd, Block) ->
    {ok, <<1, HeaderLen:32/integer, RestBlock/binary>>} =
        file:pread(Fd, Block * ?SIZE_BLOCK, ?SIZE_BLOCK),
    load_header(Fd, Block * ?SIZE_BLOCK, HeaderLen, RestBlock).


load_header(Fd, Pos, HeaderLen) ->
    load_header(Fd, Pos, HeaderLen, <<>>).


load_header(Fd, Pos, HeaderLen, RestBlock)
    TotalBytes = calculate_total_read_len(?PREFIX_SIZE, HeaderLen),
    RawBin = case TotalBytes =< byte_size(RestBlock) of
        true->
            <<RawBin0:TotalBytes/binary, _/binary>> = RestBlock,
            RawBin0;
        false ->
            ReadStart = Pos + ?PREFIX_SIZE + byte_size(RestBlock),
            ReadLen = TotalBytes - byte_size(RestBlock),
            {ok, Missing} = file:pread(Fd, ReadStart, ReadLen),
            <<RestBlock/binary, Missing/binary>>
    end,
    <<Md5Sig:16/binary, HeaderBin/binary>> =
        iolist_to_binary(remove_block_prefixes(?PREFIX_SIZE, RawBin)),
    Md5Sig = couch_crypto:hash(md5, HeaderBin),
    {ok, HeaderBin}.
```

Also, for positioning, its fairly standard Erlang style to have the same function name but varying numbers of arguments to be located next to each other in the module since they're usually doing quite similar things. Like in this case they're both loading headers. Having them separate means you have to remember when reading the module that they do slightly different things and which arity version is which. If you think you need two different behaviors in two different places it'd be better to rename them to be more descriptive of their behavior. However this case can be simplified by having one function that services both uses with minor small changes. And they should be all in the same spot in the module since they're all related.